osu.Game/Database/EFToRealmMigrator.cs

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Linq;
+using Microsoft.EntityFrameworkCore;
+using osu.Game.Configuration;
+using osu.Game.Models;
+using osu.Game.Skinning;
+
+#nullable enable
+
+namespace osu.Game.Database
+{
+    internal class EFToRealmMigrator
+    {
+        private readonly DatabaseContextFactory efContextFactory;
+        private readonly RealmContextFactory realmContextFactory;
+        private readonly OsuConfigManager config;
+
+        public EFToRealmMigrator(DatabaseContextFactory efContextFactory, RealmContextFactory realmContextFactory, OsuConfigManager config)
+        {
+            this.efContextFactory = efContextFactory;
+            this.realmContextFactory = realmContextFactory;
+            this.config = config;
+        }
+
+        public void Run()
+        {
+            using (var db = efContextFactory.GetForWrite())
+            {
+                migrateSettings(db);
+                migrateSkins(db);
+            }
+        }
+
+        private void migrateSkins(DatabaseWriteUsage db)
+        {
+            var userSkinChoice = config.GetBindable<string>(OsuSetting.Skin);
+            int.TryParse(userSkinChoice.Value, out int userSkinInt);
+
+            switch (userSkinInt)
+            {
+                case EFSkinInfo.DEFAULT_SKIN:
+                    userSkinChoice.Value = SkinInfo.DEFAULT_SKIN.ToString();
+                    break;
+
+                case EFSkinInfo.CLASSIC_SKIN:
+                    userSkinChoice.Value = SkinInfo.CLASSIC_SKIN.ToString();
+                    break;
+            }
+
+            // migrate ruleset settings. can be removed 20220530.
+            var existingSkins = db.Context.SkinInfo
+                                  .Include(s => s.Files)
+                                  .ThenInclude(f => f.FileInfo)
+                                  .ToList();
+
+            // previous entries in EF are removed post migration.
+            if (!existingSkins.Any())
+                return;
+
+            using (var realm = realmContextFactory.CreateContext())
+            using (var transaction = realm.BeginWrite())
+            {
+                // only migrate data if the realm database is empty.
+                if (!realm.All<SkinInfo>().Any(s => !s.Protected))
```

Comment:
I've actually reverted this change in 8bef50cb, because I was getting the following error on migration otherwise:

```
[runtime] 2021-12-01 18:33:48 [error]: System.NotSupportedException: The method 'All' is not supported
[runtime] 2021-12-01 18:33:48 [error]: at Realms.RealmResultsVisitor.VisitMethodCall(MethodCallExpression node)
[runtime] 2021-12-01 18:33:48 [error]: at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
```

Unfortunate, but alas...

Reply:
Could be resolved by using the `IEnumerable` version instead. Probably fine to do so here as performance is not of concern.

Reply:
Here's the diff, will apply if you agree it's better:

```diff diff --git a/osu.Game/Database/EFToRealmMigrator.cs b/osu.Game/Database/EFToRealmMigrator.cs
index 3790dc8ae9..83b05da307 100644
--- a/osu.Game/Database/EFToRealmMigrator.cs
+++ b/osu.Game/Database/EFToRealmMigrator.cs
@@ -63,8 +63,7 @@ private void migrateSkins(DatabaseWriteUsage db)
             using (var transaction = realm.BeginWrite())
             {
                 // only migrate data if the realm database is empty.
-                // note that this cannot be written as: `realm.All<SkinInfo>().All(s => s.Protected)`, because realm does not support `.All()`.
-                if (!realm.All<SkinInfo>().Any(s => !s.Protected))
+                if (realm.All<SkinInfo>().AsEnumerable().All(s => s.Protected))
                 {
                     foreach (var skin in existingSkins)
                     {
```

Reply:
Definitely clearer. I do however think that performance could matter here, since that code will run on every startup, and with the comment that I added on top of the revert either works for me, really.

Reply:
It actually doesn't because there's a pre-check to all of this that first checks if there are any EF skins left behind. And they are removed at the end of the migration process.

Reply:
Right, I forgot. But as stated above I'm not really all that fussed either way.